components/feature/recents/src/main/java/mozilla/components/feature/recents/RecentAppsStorage.kt

```diff
@@ -0,0 +1,76 @@
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package mozilla.components.feature.recents
+
+import android.content.Context
+import androidx.annotation.VisibleForTesting
+import mozilla.components.feature.recents.adapter.RecentAppAdapter
+import mozilla.components.feature.recents.db.RecentAppEntity
+import mozilla.components.feature.recents.db.RecentAppsDatabase
+
+class RecentAppsStorage(context: Context) {
+
+    companion object {
+        @VisibleForTesting
+        internal const val DEFAULT_COUNT = 0.0
+        private const val DECAY_MULTIPLIER = 0.95
+    }
+
+    internal var database: Lazy<RecentAppsDatabase> = lazy { RecentAppsDatabase.get(context) }
+
+    /**
+     * Increment the value stored in the database for the selected app. Then, apply a decay to all
+     * other apps in the database. We do not need to handle overflow as it's not reasonably expected
+     * to reach Double.MAX_VALUE for users
+     */
+    fun updateRecentApp(selectedAppName: String) {
+        val recentAppEntity = database.value.recentAppsDao().getRecentApp(selectedAppName)
+        recentAppEntity?.let {
+            it.score += 1
+            database.value.recentAppsDao().updateRecentApp(it)
+            applyDecayToRecentAppsExceptFor(it.name)
+        }
+    }
+
+    /**
+     * Get a descending ordered list of the most recent apps
+     * @param limit - size of list
+     */
+    fun getRecentAppsUpTo(limit: Int): List<RecentApp> {
+        return database.value.recentAppsDao().getRecentAppsUpTo(limit).map { entity ->
+            RecentAppAdapter(entity)
+        }
+    }
+
+    /**
+     * If there are apps that could resolve our share and are not added in our database, we add them
+     * with a 0 count, so they can be updated later when a user uses that app
+     */
+    fun updateDatabaseWithNewApps(appNames: List<String>) {
+        for (appName in appNames) {
+            if (database.value.recentAppsDao().getRecentApp(appName) == null) {
+                val recentAppEntity = RecentAppEntity(
+                        name = appName,
+                        score = DEFAULT_COUNT
+                )
+                database.value.recentAppsDao().insertRecentApp(recentAppEntity)
+            }
+        }
+    }
+
+    /**
+     * Decreases the score of the apps as they become older. This allows newly installed apps to catch
+     * up and appear in the most recent section faster.
+     */
+    private fun applyDecayToRecentAppsExceptFor(selectedAppName: String) {
+        val recentApps = database.value.recentAppsDao().getAllRecentApps()
+        for (app in recentApps) {
+            if (app.name != selectedAppName && app.score != DEFAULT_COUNT) {
+                app.score *= DECAY_MULTIPLIER
+                database.value.recentAppsDao().updateRecentApp(app)
+            }
+        }
```

Comment:
Is this something we could express in a single SQL `UPDATE` statement instead of looping over all apps? 🤔

Reply:
Instead of updating each one in a loop, I update the array now and then let Room update the entire db.